src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs

```diff
@@ -359,36 +350,39 @@ public async Task<ImmutableArray<Diagnostic>> GetAnalyzerCompilationDiagnosticsA
-            var analysisScope = new AnalysisScope(_compilation, _analysisOptions.Options, analyzers, hasAllAnalyzers, concurrentAnalysis: _analysisOptions.ConcurrentAnalysis, categorizeDiagnostics: true);
+            var analysisScope = _compilationAnalysisScope.WithAnalyzers(analyzers, hasAllAnalyzers);
             await ComputeAnalyzerDiagnosticsAsync(analysisScope, cancellationToken).ConfigureAwait(false);
             return _analysisResultBuilder.ToAnalysisResult(analyzers, analysisScope, cancellationToken);
         }
 
         private async Task<ImmutableArray<Diagnostic>> GetAnalyzerDiagnosticsCoreAsync(ImmutableArray<DiagnosticAnalyzer> analyzers, CancellationToken cancellationToken)
         {
             var hasAllAnalyzers = analyzers.Length == Analyzers.Length;
-            var analysisScope = new AnalysisScope(_compilation, _analysisOptions.Options, analyzers, hasAllAnalyzers, concurrentAnalysis: _analysisOptions.ConcurrentAnalysis, categorizeDiagnostics: true);
+            var analysisScope = _compilationAnalysisScope.WithAnalyzers(analyzers, hasAllAnalyzers);
             await ComputeAnalyzerDiagnosticsAsync(analysisScope, cancellationToken).ConfigureAwait(false);
             return _analysisResultBuilder.GetDiagnostics(analysisScope, getLocalDiagnostics: true, getNonLocalDiagnostics: true);
         }
 
-        private static AnalyzerDriver CreateDriverForComputingDiagnosticsWithoutStateTracking(Compilation compilation, ImmutableArray<DiagnosticAnalyzer> analyzers)
+        private static async Task<AnalyzerDriver> CreateAndInitializeDriverAsync(
+            Compilation compilation,
+            CompilationWithAnalyzersOptions analysisOptions,
+            AnalysisScope analysisScope,
+            bool categorizeDiagnostics,
+            CancellationToken cancellationToken)
         {
-            // Create and attach a new driver instance to compilation, do not used a pooled instance from '_driverPool'.
-            // Additionally, use a new AnalyzerManager and don't use the analyzer manager instance
-            // stored onto the field '_analyzerManager'.
-            // Pooled _driverPool and shared _analyzerManager is used for partial/state-based analysis, and
-            // we want to compute diagnostics without any state tracking here.
-
-            return compilation.CreateAnalyzerDriver(analyzers, new AnalyzerManager(analyzers), severityFilter: SeverityFilter.None);
+            var driver = compilation.CreateAnalyzerDriver(analysisScope.Analyzers, new AnalyzerManager(analysisScope.Analyzers), severityFilter: SeverityFilter.None);
+            driver.Initialize(compilation, analysisOptions, new CompilationData(compilation), analysisScope, categorizeDiagnostics, trackSuppressedDiagnosticIds: false, cancellationToken);
+            cancellationToken.ThrowIfCancellationRequested();
+            await driver.WhenInitializedTask.ConfigureAwait(false);
```

Comment:
odd pattern... why not have driver.Initialize just be async itself?

Reply:
This pattern is due to the fact that for batch compile scenario we don't want to await the initialization task right after creation. We do the await when we first start processing the compilation events generated from compiling.